File: python_files/raylib_redo/breakout.py
import pyray as p

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "Breakout"
SCREEN_BACKGROUND: p.Color = p.RAYWHITE

ROWS: int = 5
COLS: int = 10
BRICK_WIDTH: int = 79
BRICK_HEIGHT: int = 30
BRICK_GAP: int = 2


class Bricks:
    def __init__(self) -> None:
        self.list: list[p.Vector2] = []
        self.color: p.Color = p.GRAY

        for i in range(ROWS):
            for j in range(COLS):
                x: int = j * (BRICK_WIDTH + BRICK_GAP)
                y: int = i * (BRICK_HEIGHT + BRICK_GAP)

                self.list.append(p.Vector2(x, y))

    def draw(self) -> None:
        for brick in self.list:
            p.draw_rectangle_rec(
                p.Rectangle(brick.x, brick.y, BRICK_WIDTH, BRICK_HEIGHT), self.color
            )


class Paddle:
    def __init__(self) -> None:
        self.width: float = 100.0
        self.height: float = 10.0
        self.initial_x: float = p.get_screen_width() / 2 - self.width / 2
        self.initial_y: float = p.get_screen_height() - self.height - 10.0
        self.x: float = self.initial_x
        self.y: float = self.initial_y
        self.speed: float = 300.0
        self.color: p.Color = p.BLACK

    def reset(self) -> None:
        self.x = self.initial_x
        self.y = self.initial_y

    def draw(self) -> None:
        p.draw_rectangle_rec(
            p.Rectangle(self.x, self.y, self.width, self.height), self.color
        )

    def update(self) -> None:
        # move paddle
        if p.is_key_down(p.KeyboardKey.KEY_LEFT) and self.x > 0:
            self.x -= self.speed * p.get_frame_time()
        if (
            p.is_key_down(p.KeyboardKey.KEY_RIGHT)
            and self.x < p.get_screen_width() - self.width
        ):
            self.x += self.speed * p.get_frame_time()


class Ball:
    def __init__(self, paddle: Paddle) -> None:
        self.paddle: Paddle = paddle
        self.radius: float = 10.0
        self.initial_x: float = self.paddle.x + self.paddle.width / 2
        self.initial_y: float = self.paddle.y - self.paddle.height - (self.radius * 2)
        self.x: float = self.initial_x
        self.y: float = self.initial_y
        self.speed: float = 400.0
        self.direction: p.Vector2 = p.Vector2(
            -1 if p.get_random_value(0, 1) == 0 else 1,
            -1 if p.get_random_value(0, 1) == 0 else 1,
        )
        self.color: p.Color = p.RED
        self.active: bool = False

    def reset(self) -> None:
        self.x = self.initial_x
        self.y = self.initial_y
        self.direction = p.Vector2(
            -1 if p.get_random_value(0, 1) == 0 else 1,
            -1 if p.get_random_value(0, 1) == 0 else 1,
        )
        self.active = False

    def draw(self) -> None:
        p.draw_circle_v(p.Vector2(self.x, self.y), self.radius, self.color)

    def update(self) -> None:
        # update ball position
        if not self.active:
            self.x: float = self.paddle.x + self.paddle.width / 2
            self.y: float = self.paddle.y - self.paddle.height - (self.radius * 2)

        # activate ball
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.active = not self.active

        # move ball
        if self.active:
            self.x += self.direction.x * self.speed * p.get_frame_time()
            self.y += self.direction.y * self.speed * p.get_frame_time()

        # normalize direction vector
        if self.direction.x != 0 and self.direction.y != 0:
            self.direction = p.vector2_normalize(self.direction)

        # ball bounds
        if self.x < self.radius or self.x > p.get_screen_width() - self.radius:
            self.direction.x *= -1
        if self.y < self.radius:
            self.direction.y *= -1


class Game:
    def __init__(self) -> None:
        self.lives: int = 5
        self.game_over: bool = False
        self.game_won: bool = False

        self.bricks: Bricks = Bricks()
        self.paddle: Paddle = Paddle()
        self.ball: Ball = Ball(self.paddle)

    def reset(self) -> None:
        self.paddle.reset()
        self.ball.reset()

    def center_and_draw_text(
        self, text: str, font_size: float, rect: p.Rectangle, tint: p.Color
    ) -> None:
        font: p.Font = p.get_font_default()
        text: str = text
        font_size: float = font_size
        text_width: float = p.measure_text(text, int(font_size))
        text_x: float = rect.x + rect.width / 2 - text_width / 2
        text_y: float = rect.y + rect.height / 2 - font_size / 2
        spacing: float = 5.0
        tint: p.Color = tint

        p.draw_text_ex(font, text, p.Vector2(text_x, text_y), font_size, spacing, tint)

    def draw(self) -> None:
        # draw lives
        p.draw_text(str(self.lives), 20, p.get_screen_height() - 50, 40, p.BLACK)

        self.ball.draw()
        self.paddle.draw()
        self.bricks.draw()

        if self.game_over:
            self.center_and_draw_text(
                "GAME OVER",
                40.0,
                p.Rectangle(0, 0, p.get_screen_width(), p.get_screen_height()),
                p.BLACK,
            )

        if self.game_won:
            self.center_and_draw_text(
                "YOU WIN",
                40.0,
                p.Rectangle(0, 0, p.get_screen_width(), p.get_screen_height()),
                p.BLACK,
            )

    def update(self) -> None:
        if not self.game_won and not self.game_over:
            # game over condition:
            if self.lives <= 0:
                self.game_over = True

            # game won condition
            if len(self.bricks.list) == 0:
                self.game_won = True

            self.paddle.update()
            self.ball.update()

            # ball collision paddle
            if p.check_collision_circle_rec(
                p.Vector2(self.ball.x, self.ball.y),
                self.ball.radius,
                p.Rectangle(
                    self.paddle.x, self.paddle.y, self.paddle.width, self.paddle.height
                ),
            ):
                self.ball.direction.y *= -1

            # ball collision bricks
            for brick in self.bricks.list[:]:
                if p.check_collision_circle_rec(
                    p.Vector2(self.ball.x, self.ball.y),
                    self.ball.radius,
                    p.Rectangle(brick.x, brick.y, BRICK_WIDTH, BRICK_HEIGHT),
                ):
                    self.ball.direction.y *= -1
                    self.bricks.list.remove(brick)

            # update lives
            if self.ball.y > p.get_screen_height():
                self.lives -= 1
                self.reset()
        else:
            if p.is_key_down(p.KeyboardKey.KEY_ENTER):
                self.game_over = False
                self.game_won = False
                self.lives = 5

                self.bricks = Bricks()
                self.paddle = Paddle()
                self.ball = Ball(self.paddle)


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    game: Game = Game()

    while not p.window_should_close():
        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        game.draw()
        game.update()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: c_files/raylib_redo/breakout.c
#include <raylib.h>
#include <raymath.h>
#include <stdbool.h>

#define ROWS 5
#define COLS 10
#define BRICK_WIDTH 79
#define BRICK_HEIGHT 30
#define BRICK_GAP 2

// BRICK

typedef struct Brick
{
  float x;
  float y;
  float width;
  float height;
  Color color;
  bool is_active;

} Brick;

void
brick_init(Brick* brick, float x, float y)
{
  brick->x = x;
  brick->y = y;
  brick->width = BRICK_WIDTH;
  brick->height = BRICK_HEIGHT;
  brick->color = GRAY;
  brick->is_active = true;
}

void
brick_draw(Brick* brick)
{
  if (brick->is_active) {
    DrawRectangleRec(
      (Rectangle){ brick->x, brick->y, brick->width, brick->height },
      brick->color);
  }
}

// PADDLE

typedef struct Paddle
{
  float initial_x;
  float intiial_y;
  float x;
  float y;
  float width;
  float height;
  float speed;
  Color color;

} Paddle;

void
paddle_init(Paddle* paddle)
{
  paddle->width = 100.0f;
  paddle->height = 10.0f;
  paddle->initial_x = GetScreenWidth() * 0.5f - paddle->width / 2;
  paddle->intiial_y = GetScreenHeight() - paddle->height - 10.0f;
  paddle->x = paddle->initial_x;
  paddle->y = paddle->intiial_y;
  paddle->speed = 300.0f;
  paddle->color = BLACK;
}

void
paddle_reset(Paddle* paddle)
{
  paddle->x = paddle->initial_x;
  paddle->y = paddle->intiial_y;
}

void
paddle_draw(Paddle* paddle)
{
  DrawRectangleRec(
    (Rectangle){ paddle->x, paddle->y, paddle->width, paddle->height },
    paddle->color);
}

void
paddle_update(Paddle* paddle)
{
  // move paddle
  if (IsKeyDown(KEY_LEFT) && paddle->x > 0) {
    paddle->x -= paddle->speed * GetFrameTime();
  }
  if (IsKeyDown(KEY_RIGHT) && paddle->x < GetScreenWidth() - paddle->width) {
    paddle->x += paddle->speed * GetFrameTime();
  }
}

// BALL

typedef struct Ball
{
  float initial_x;
  float initial_y;
  float x;
  float y;
  float radius;
  float speed;
  Vector2 direction;
  Color color;
  bool is_active;

} Ball;

void
ball_init(Ball* ball, Paddle* paddle)
{
  ball->radius = 10.0f;
  ball->initial_x = paddle->x + paddle->width / 2;
  ball->initial_y = paddle->y - ball->radius * 2;
  ball->x = ball->initial_x;
  ball->y = ball->initial_y;
  ball->speed = 400.0f;
  ball->direction = (Vector2){
    GetRandomValue(0, 1) == 0 ? -1 : 1,
    GetRandomValue(0, 1) == 0 ? -1 : 1,
  };
  ball->color = RED;
  ball->is_active = false;
}

void
ball_reset(Ball* ball)
{
  ball->x = ball->initial_x;
  ball->y = ball->initial_y;
  ball->direction = (Vector2){
    GetRandomValue(0, 1) == 0 ? -1 : 1,
    GetRandomValue(0, 1) == 0 ? -1 : 1,
  };
  ball->is_active = false;
}

void
ball_draw(Ball* ball)
{
  DrawCircleV((Vector2){ ball->x, ball->y }, ball->radius, ball->color);
}

void
ball_update(Ball* ball, Paddle* paddle)
{
  // update ball position
  if (!ball->is_active) {
    ball->x = paddle->x + paddle->width / 2;
    ball->y = paddle->y - ball->radius * 2;
  }

  // activate ball
  if (IsKeyPressed(KEY_SPACE)) {
    ball->is_active = !ball->is_active;
  }

  // move ball
  if (ball->is_active) {
    ball->x += ball->direction.x * ball->speed * GetFrameTime();
    ball->y += ball->direction.y * ball->speed * GetFrameTime();
  }

  // normalize direction vector
  if (ball->direction.x != 0 && ball->direction.y != 0) {
    ball->direction = Vector2Normalize(ball->direction);
  }

  // ball bounds
  if (ball->x < ball->radius || ball->x > GetScreenWidth() - ball->radius) {
    ball->direction.x *= -1;
  }
  if (ball->y < ball->radius) {
    ball->direction.y *= -1;
  }
}

// UTILITY FUNCTIONS

void
reset(Paddle* paddle, Ball* ball)
{
  ball_reset(ball);
  paddle_reset(paddle);
}

void
center_and_draw_text(char* text,
                     int font_size,
                     int x,
                     int y,
                     int width,
                     int height)
{
  int text_width = MeasureText(text, font_size);
  int text_x = x + width / 2 - text_width / 2;
  int text_y = y + height / 2 - font_size / 2;
  Color color = BLACK;
  DrawText(text, text_x, text_y, font_size, color);
}

// MAIN

int
main(void)
{
  const int SCREEN_WIDTH = 800;
  const int SCREEN_HEIGHT = 600;
  const char SCREEN_TITLE[] = "Breakout";
  const Color SCREEN_BACKGROUND = RAYWHITE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  int lives = 5;
  bool game_over = false;
  bool game_won = false;

  Brick bricks[ROWS * COLS];

  for (int i = 0; i < ROWS; i++) {
    for (int j = 0; j < COLS; j++) {
      float x = j * (BRICK_WIDTH + BRICK_GAP);
      float y = i * (BRICK_HEIGHT + BRICK_GAP);

      Brick brick;
      brick_init(&brick, x, y);

      bricks[i * COLS + j] = brick;
    }
  }

  Paddle paddle;
  Ball ball;

  paddle_init(&paddle);
  ball_init(&ball, &paddle);

  while (!WindowShouldClose()) {

    // UPDATE

    // game over condition
    if (lives <= 0) {
      game_over = true;
    }

    // game won condition
    bool bricks_active = false;
    for (int i = 0; i < ROWS * COLS; i++) {
      if (bricks[i].is_active) {
        bricks_active = true;
        break;
      }
    }

    if (!bricks_active) {
      game_won = true;
    }

    if (!game_over && !game_won) {
      paddle_update(&paddle);
      ball_update(&ball, &paddle);

      // ball collision paddle
      if (CheckCollisionCircleRec(
            (Vector2){ ball.x, ball.y },
            ball.radius,
            (Rectangle){ paddle.x, paddle.y, paddle.width, paddle.height })) {
        ball.direction.y *= -1;
      }

      // ball collision bricks
      for (int i = 0; i < ROWS * COLS; i++) {
        if (CheckCollisionCircleRec((Vector2){ ball.x, ball.y },
                                    ball.radius,
                                    (Rectangle){ bricks[i].x,
                                                 bricks[i].y,
                                                 bricks[i].width,
                                                 bricks[i].height })) {
          if (bricks[i].is_active) {
            ball.direction.y *= -1;
            bricks[i].is_active = false;
          }
        }
      }

      // update lives
      if (ball.y > GetScreenHeight() - ball.radius) {
        lives--;
        reset(&paddle, &ball);
      }

    } else {
      if (IsKeyPressed(KEY_ENTER)) {
        lives = 5;
        game_over = false;
        game_won = false;

        paddle_init(&paddle);
        ball_init(&ball, &paddle);
      }
    }

    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);

    // DRAW

    DrawText(TextFormat("%d", lives), 20, GetScreenHeight() - 50, 40, BLACK);

    for (int i = 0; i < ROWS * COLS; i++) {
      brick_draw(&bricks[i]);
    }

    paddle_draw(&paddle);
    ball_draw(&ball);

    if (game_over) {
      center_and_draw_text(
        "GAME OVER", 40, 0, 0, (int)GetScreenWidth(), (int)GetScreenHeight());
    }
    if (game_won) {
      center_and_draw_text(
        "YOU WIN", 40, 0, 0, (int)GetScreenWidth(), (int)GetScreenHeight());
    }

    EndDrawing();
  }

  CloseWindow();

  return 0;
}
